File: src/name.cpp
/// \file name.cpp
/// \brief Implementation of ida::name — naming, demangling, properties.

#include "detail/sdk_bridge.hpp"
#include <ida/name.hpp>
#include <demangle.hpp>

namespace ida::name {

namespace {

bool in_range(Address address, const ListOptions& options) {
    if (options.start != BadAddress && address < options.start)
        return false;
    if (options.end != BadAddress && address >= options.end)
        return false;
    return true;
}

bool include_name(bool user_defined, const ListOptions& options) {
    if (user_defined)
        return options.include_user_defined;
    return options.include_auto_generated;
}

} // anonymous namespace

// ── Core naming ─────────────────────────────────────────────────────────

Status set(Address ea, std::string_view name) {
    qstring qn = ida::detail::to_qstring(name);
    if (!set_name(ea, qn.c_str(), SN_NOCHECK))
        return std::unexpected(Error::sdk("set_name failed", std::to_string(ea)));
    return ida::ok();
}

Status force_set(Address ea, std::string_view name) {
    qstring qn = ida::detail::to_qstring(name);
    if (!set_name(ea, qn.c_str(), SN_FORCE))
        return std::unexpected(Error::sdk("force set_name failed", std::to_string(ea)));
    return ida::ok();
}

Status remove(Address ea) {
    if (!set_name(ea, "", 0))
        return std::unexpected(Error::sdk("remove name failed", std::to_string(ea)));
    return ida::ok();
}

Result<std::string> get(Address ea) {
    qstring qn = get_name(ea);
    if (qn.empty())
        return std::unexpected(Error::not_found("No name at address", std::to_string(ea)));
    return ida::detail::to_string(qn);
}

Result<std::string> demangled(Address ea, DemangleForm form) {
    // Map our form enum to GN_ flags.
    int gtn_flags = GN_DEMANGLED;
    switch (form) {
        case DemangleForm::Short:
            gtn_flags |= GN_SHORT;
            break;
        case DemangleForm::Long:
            // Default demangled output.
            break;
        case DemangleForm::Full:
            gtn_flags |= GN_LONG;
            break;
    }

    qstring qn;
    if (get_ea_name(&qn, ea, gtn_flags) <= 0)
        return std::unexpected(Error::not_found("No demangled name at address",
                                                std::to_string(ea)));
    return ida::detail::to_string(qn);
}

Result<std::string> demangled(std::string_view symbol, DemangleForm form) {
    if (symbol.empty())
        return std::unexpected(Error::validation("Symbol cannot be empty"));
    if (symbol.find('\0') != std::string_view::npos) {
        return std::unexpected(Error::validation(
            "Symbol contains an embedded NUL"));
    }

    uint32 disable_mask = 0;
    switch (form) {
        case DemangleForm::Short:
            disable_mask = MNG_SHORT_FORM;
            break;
        case DemangleForm::Long:
            disable_mask = MNG_LONG_FORM;
            break;
        case DemangleForm::Full:
            break;
    }

    const qstring input = ida::detail::to_qstring(symbol);
    qstring output;
    if (::demangle_name(&output, input.c_str(), disable_mask, DQT_FULL) <= 0
        || output.empty()) {
        return std::unexpected(Error::not_found(
            "Symbol could not be demangled",
            std::string(symbol)));
    }
    return ida::detail::to_string(output);
}

Result<Address> resolve(std::string_view name, Address context) {
    qstring qn = ida::detail::to_qstring(name);
    ea_t from = (context == BadAddress) ? BADADDR : static_cast<ea_t>(context);
    ea_t result = get_name_ea(from, qn.c_str());
    if (!ida::detail::is_valid(result))
        return std::unexpected(Error::not_found("Name not resolved",
                                                std::string(name)));
    return static_cast<Address>(result);
}

// ── Name inventory ───────────────────────────────────────────────────────

Result<std::vector<Entry>> all(const ListOptions& options) {
    if (!options.include_user_defined && !options.include_auto_generated) {
        return std::unexpected(Error::validation(
            "At least one name category must be included"));
    }
    if (options.start != BadAddress
        && options.end != BadAddress
        && options.start > options.end) {
        return std::unexpected(Error::validation(
            "Invalid name range: start is greater than end",
            std::to_string(options.start) + ":" + std::to_string(options.end)));
    }

    const std::size_t total = get_nlist_size();
    std::vector<Entry> out;
    out.reserve(total);

    for (std::size_t index = 0; index < total; ++index) {
        const ea_t ea = get_nlist_ea(index);
        if (!ida::detail::is_valid(ea))
            continue;

        const Address address = static_cast<Address>(ea);
        if (!in_range(address, options))
            continue;

        const char* raw_name = get_nlist_name(index);
        if (raw_name == nullptr || *raw_name == '\0')
            continue;

        const flags64_t flags = get_flags(ea);
        const bool user_defined = flags != 0 && has_user_name(flags);
        if (!include_name(user_defined, options))
            continue;

        Entry entry;
        entry.address = address;
        entry.name = raw_name;
        entry.user_defined = user_defined;
        entry.auto_generated = !user_defined;
        out.push_back(std::move(entry));
    }

    return out;
}

Result<std::vector<Entry>> all_user_defined(Address start, Address end) {
    ListOptions options;
    options.start = start;
    options.end = end;
    options.include_user_defined = true;
    options.include_auto_generated = false;
    return all(options);
}

// ── Name properties ─────────────────────────────────────────────────────

bool is_public(Address ea) {
    return is_public_name(ea);
}

bool is_weak(Address ea) {
    return is_weak_name(ea);
}

bool is_user_defined(Address ea) {
    flags64_t f = get_flags(ea);
    if (f == 0)
        return false;
    return has_any_name(f) && has_user_name(f);
}

bool is_auto_generated(Address ea) {
    // A name is auto-generated if it exists but is not user-defined.
    // SDK has has_user_name() which checks if the name was set by the user.
    flags64_t f = get_flags(ea);
    if (f == 0) return false;
    return has_any_name(f) && !has_user_name(f);
}

Result<bool> is_valid_identifier(std::string_view text) {
    if (text.empty())
        return false;
    std::string value(text);
    return ::is_uname(value.c_str());
}

Result<std::string> sanitize_identifier(std::string_view text) {
    if (text.empty())
        return std::unexpected(Error::validation("Identifier cannot be empty"));

    qstring value = ida::detail::to_qstring(text);
    if (!validate_name(&value, VNT_IDENT, SN_NOCHECK)) {
        return std::unexpected(Error::validation("Identifier cannot be sanitized",
                                                 std::string(text)));
    }
    return ida::detail::to_string(value);
}

Status set_public(Address ea, bool value) {
    if (value)
        make_name_public(ea);
    else
        make_name_non_public(ea);
    return ida::ok();
}

Status set_weak(Address ea, bool value) {
    if (value)
        make_name_weak(ea);
    else
        make_name_non_weak(ea);
    return ida::ok();
}

} // namespace ida::name

File: bindings/rust/idax/src/name.rs
//! Naming, demangling, and name property operations.
//!
//! Mirrors the C++ `ida::name` namespace.

use crate::address::{Address, BAD_ADDRESS};
use crate::error::{self, Error, Result, Status};
use std::ffi::CString;

/// Demangle form.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(i32)]
pub enum DemangleForm {
    Short = 0,
    Long = 1,
    Full = 2,
}

/// Enumerated name entry.
#[derive(Debug, Clone)]
pub struct Entry {
    pub address: Address,
    pub name: String,
    pub user_defined: bool,
    pub auto_generated: bool,
}

/// Options for name inventory enumeration.
#[derive(Debug, Clone)]
pub struct ListOptions {
    pub start: Address,
    pub end: Address,
    pub include_user_defined: bool,
    pub include_auto_generated: bool,
}

impl Default for ListOptions {
    fn default() -> Self {
        Self {
            start: BAD_ADDRESS,
            end: BAD_ADDRESS,
            include_user_defined: true,
            include_auto_generated: true,
        }
    }
}

// ---------------------------------------------------------------------------
// Core naming
// ---------------------------------------------------------------------------

/// Set or replace the name at `address`.
pub fn set(address: Address, name: &str) -> Status {
    let c_name = CString::new(name).map_err(|_| Error::validation("invalid name"))?;
    let ret = unsafe { idax_sys::idax_name_set(address, c_name.as_ptr()) };
    error::int_to_status(ret, "name::set failed")
}

/// Force-set a name, appending a numeric suffix if the name is taken.
pub fn force_set(address: Address, name: &str) -> Status {
    let c_name = CString::new(name).map_err(|_| Error::validation("invalid name"))?;
    let ret = unsafe { idax_sys::idax_name_force_set(address, c_name.as_ptr()) };
    error::int_to_status(ret, "name::force_set failed")
}

/// Remove the name at `address`.
pub fn remove(address: Address) -> Status {
    let ret = unsafe { idax_sys::idax_name_remove(address) };
    error::int_to_status(ret, "name::remove failed")
}

/// Get the name at `address`.
pub fn get(address: Address) -> Result<String> {
    unsafe {
        let mut out: *mut std::ffi::c_char = std::ptr::null_mut();
        let ret = idax_sys::idax_name_get(address, &mut out);
        if ret != 0 {
            return Err(error::consume_last_error("name::get failed"));
        }
        error::cstr_to_string_free(out, "name::get returned null")
    }
}

/// Get the demangled name at `address`.
pub fn demangled(address: Address, form: DemangleForm) -> Result<String> {
    unsafe {
        let mut out: *mut std::ffi::c_char = std::ptr::null_mut();
        let ret = idax_sys::idax_name_demangled(address, form as i32, &mut out);
        if ret != 0 {
            return Err(error::consume_last_error("name::demangled failed"));
        }
        error::cstr_to_string_free(out, "name::demangled returned null")
    }
}

/// Demangle an arbitrary mangled symbol without requiring a database address.
pub fn demangle(symbol: &str, form: DemangleForm) -> Result<String> {
    let symbol =
        CString::new(symbol).map_err(|_| Error::validation("symbol contains an embedded NUL"))?;
    unsafe {
        let mut out: *mut std::ffi::c_char = std::ptr::null_mut();
        let ret = idax_sys::idax_name_demangle(symbol.as_ptr(), form as i32, &mut out);
        if ret != 0 {
            return Err(error::consume_last_error("name::demangle failed"));
        }
        error::cstr_to_string_free(out, "name::demangle returned null")
    }
}

/// Resolve a name to an address.
pub fn resolve(name: &str, context: Address) -> Result<Address> {
    let c_name = CString::new(name).map_err(|_| Error::validation("invalid name"))?;
    let mut out: Address = BAD_ADDRESS;
    let ret = unsafe { idax_sys::idax_name_resolve(c_name.as_ptr(), context, &mut out) };
    if ret != 0 {
        Err(error::consume_last_error("name::resolve failed"))
    } else {
        Ok(out)
    }
}

unsafe fn entries_from_ffi(entries_ptr: *mut idax_sys::IdaxNameEntry, count: usize) -> Vec<Entry> {
    if entries_ptr.is_null() || count == 0 {
        return Vec::new();
    }
    let mut out = Vec::with_capacity(count);
    let entries = unsafe { std::slice::from_raw_parts(entries_ptr, count) };
    for entry in entries {
        out.push(Entry {
            address: entry.address,
            name: if entry.name.is_null() {
                String::new()
            } else {
                unsafe {
                    std::ffi::CStr::from_ptr(entry.name)
                        .to_string_lossy()
                        .into_owned()
                }
            },
            user_defined: entry.user_defined != 0,
            auto_generated: entry.auto_generated != 0,
        });
    }
    unsafe { idax_sys::idax_name_entries_free(entries_ptr, count) };
    out
}

/// Enumerate names using explicit range and origin filters.
pub fn all(options: &ListOptions) -> Result<Vec<Entry>> {
    unsafe {
        let mut entries_ptr: *mut idax_sys::IdaxNameEntry = std::ptr::null_mut();
        let mut count = 0usize;
        let ret = idax_sys::idax_name_all(
            options.start,
            options.end,
            options.include_user_defined as i32,
            options.include_auto_generated as i32,
            &mut entries_ptr,
            &mut count,
        );
        if ret != 0 {
            return Err(error::consume_last_error("name::all failed"));
        }
        Ok(entries_from_ffi(entries_ptr, count))
    }
}

/// Enumerate only user-defined names, optionally in `[start, end)`.
pub fn all_user_defined(start: Address, end: Address) -> Result<Vec<Entry>> {
    unsafe {
        let mut entries_ptr: *mut idax_sys::IdaxNameEntry = std::ptr::null_mut();
        let mut count: usize = 0;
        let ret = idax_sys::idax_name_all_user_defined(start, end, &mut entries_ptr, &mut count);
        if ret != 0 {
            return Err(error::consume_last_error("name::all_user_defined failed"));
        }
        Ok(entries_from_ffi(entries_ptr, count))
    }
}

// ---------------------------------------------------------------------------
// Name properties
// ---------------------------------------------------------------------------

/// Is the name at `address` public?
pub fn is_public(address: Address) -> bool {
    unsafe { idax_sys::idax_name_is_public(address) != 0 }
}

/// Is the name at `address` weak?
pub fn is_weak(address: Address) -> bool {
    unsafe { idax_sys::idax_name_is_weak(address) != 0 }
}

/// Is the name at `address` user-defined?
pub fn is_user_defined(address: Address) -> bool {
    unsafe { idax_sys::idax_name_is_user_defined(address) != 0 }
}

/// Is the name at `address` auto-generated?
pub fn is_auto_generated(address: Address) -> bool {
    unsafe { idax_sys::idax_name_is_auto_generated(address) != 0 }
}

/// Validate a user-facing identifier according to IDA naming rules.
pub fn is_valid_identifier(text: &str) -> Result<bool> {
    let c_text = CString::new(text).map_err(|_| Error::validation("invalid identifier text"))?;
    let mut out: i32 = 0;
    let ret = unsafe { idax_sys::idax_name_is_valid_identifier(c_text.as_ptr(), &mut out) };
    if ret != 0 {
        Err(error::consume_last_error(
            "name::is_valid_identifier failed",
        ))
    } else {
        Ok(out != 0)
    }
}

/// Normalize an identifier by replacing invalid characters where possible.
pub fn sanitize_identifier(text: &str) -> Result<String> {
    let c_text = CString::new(text).map_err(|_| Error::validation("invalid identifier text"))?;
    unsafe {
        let mut out: *mut std::ffi::c_char = std::ptr::null_mut();
        let ret = idax_sys::idax_name_sanitize_identifier(c_text.as_ptr(), &mut out);
        if ret != 0 {
            return Err(error::consume_last_error(
                "name::sanitize_identifier failed",
            ));
        }
        error::cstr_to_string_free(out, "name::sanitize_identifier returned null")
    }
}

/// Set or clear the public flag.
pub fn set_public(address: Address, value: bool) -> Status {
    let ret = unsafe { idax_sys::idax_name_set_public(address, value as i32) };
    error::int_to_status(ret, "name::set_public failed")
}

/// Set or clear the weak flag.
pub fn set_weak(address: Address, value: bool) -> Status {
    let ret = unsafe { idax_sys::idax_name_set_weak(address, value as i32) };
    error::int_to_status(ret, "name::set_weak failed")
}
